oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java

```diff
@@ -144,6 +144,9 @@ private JWT parse(String token) {
 		}
 		catch (Exception ex) {
 			this.logger.trace("Failed to parse token", ex);
+			if(ex.getCause() instanceof ParseException) {
+				throw new BadJwtException(String.format(DECODING_ERROR_MESSAGE_TEMPLATE, "Malformed header"), ex);
```

Comment:
I think that the body could also be malformed. Can this say "Malformed token" instead?

Reply:
It appears like the exception here is thrown by the JWTParser when it cannot decode the header.

https://bitbucket.org/connect2id/nimbus-jose-jwt/src/release-9.24.4/src/main/java/com/nimbusds/jwt/JWTParser.java/#-71

Reply:
It can also be thrown here: https://bitbucket.org/connect2id/nimbus-jose-jwt/src/d8a4358e56328637974b343a0834da67e20a6e30/src/main/java/com/nimbusds/jwt/SignedJWT.java#lines-98, which `JWTParser#parse` calls.

Further, the JavaDoc says:

```
* @throws ParseException If the string couldn't be parsed to a valid 
* unsecured, signed or encrypted JWT.
```

which I think pays dividends to follow.

Reply:
I agree, and I don't mean to go against the grain,  but the case where the body of the token is malformed seems to have already been addressed, indicated by this test case.
https://github.com/spring-projects/spring-security/blob/c336ca49fbeeea5b8acf6b87ccfcb3685358688f/oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java#L188-L195. Upon reviewing the issue submitted gh-13730, the exception that contained the unexpected message was thrown because the user introduced a malformed header, hence the reason for adding  "Malformed header" to the exception message, since a malformed token body is already addressed.

Reply:
I understand now. I agree, "malformed token" would be a more appropriate message.